### PR TITLE
Permessage-deflate extension

### PR DESCRIPTION
**Pull request by SirCipher**
This PR adds support for permessage-deflate for both inbound and outbound messages through the use of WebSocket extensions. 
I've also added this support to the tokio-tungstenite crate in a separate PR.

This is feature-gated through the `deflate` flag and uses the Flate2 crate for inflating and deflating messages.

I've updated the client and server Autobahn results and all unit tests for both pass.

### CODE REVIEW

File: Cargo.toml
Comment (application-developer-DA) [inline Cargo.toml:38]: BTW, do we want to rely on `zlib`? @agalakhov, what do you think?

File: README.md
Comment (application-developer-DA) [inline README.md:61]: Perhaps we can remove this line now ;)

File: src/protocol/mod.rs
Comment (application-developer-DA) [inline src/protocol/mod.rs:608]: If I'm not missing something, I think this could be rewritten as:

```rust
frame = self.config.encoder.on_send_frame(frame)?;
```

File: src/protocol/mod.rs
Comment (application-developer-DA) [inline src/protocol/mod.rs:534]: This seems to be the same as `self.config.encoder.on_receive_frame(frame).map_err(Into::into)`, although i'm not sure that even `map_err()` would be required.

File: src/protocol/mod.rs
Comment (application-developer-DA) [inline src/protocol/mod.rs:439]: Does it have any adverse effects that we postpone this check?

File: src/protocol/mod.rs
Comment (application-developer-DA) [inline src/protocol/mod.rs:531]: Why `_` instead of `OpCode::Control`?

File: src/handshake/client.rs
Comment (application-developer-DA) [inline src/handshake/client.rs:129]: Could it be rewritten like this?

`let request = config.map(|config| config.encoder.on_make_request(request)).unwrap_or(request);`

It would probably look cleaner like this.

File: src/handshake/client.rs
Comment (application-developer-DA) [inline src/handshake/client.rs:106]: Oh, that looks like a bold move :) Are we sure that it's always `Some`? Because if we do, then probably `config` must not be an `Option`.

File: src/handshake/client.rs
Comment (application-developer-DA) [inline src/handshake/client.rs:252]: I think it can be rewritten as:

```rust
config.map(|conf| conf.encoder.on_response(response)).transpose()?;
```

File: src/handshake/client.rs
Comment (application-developer-DA) [inline src/handshake/client.rs:71]: It looks a big counter-intuitive to me: why do we make it `Some(...)` if it's stored as an `Option`? Something is wrong with it. We either should not pass `Some` or not store it as `Option`? Or am I missing something?

File: src/handshake/server.rs
Comment (application-developer-DA) [inline src/handshake/server.rs:266]: See error handling mentioned below.

File: src/protocol/mod.rs
Comment (application-developer-DA) [inline src/protocol/mod.rs:282]: Out of curiosity: why did you decide to move out the config creation line?

File: src/extensions/uncompressed.rs
Comment (application-developer-DA) [inline src/extensions/uncompressed.rs:101]: It's a bit dangerous to panic here as we don't really know (from the compiler's point of view there is no guarantee for that) that it's really `unreachable!()`. `unreachable!()` statements are ok in some cases, when the place it's used is close to the code that ensures the invariant. But here we don't really have any guarantee that we never get here.

File: src/extensions/deflate.rs
Comment (application-developer-DA) [inline src/extensions/deflate.rs:1]: Did you write this file manually? Or is it a modified copy-paste from `deflate`? I'm asking, because it's quite a huge file and I and @agalakhov will have to validate it carefully as it contains `loop`s, `unwrap`s and some copy-pasting.

**Review by application-developer-DA — CHANGES_REQUESTED** [review]
First of all: thanks for the PR and the changes that you made!

I left some comments, which I think must be addressed before we can merge them. I also have not yet checked `deflate.rs` implementation.

Also I have one general question: do we really need an additional type parameter in all places? Maybe we can avoid it? Having a generic trait for extensions is fine, but can't we use many extensions and is a `Config` a right place to choose the extension? I'm not sure at the moment. It seems like the current implementation assumes that `encoder` is just a concrete implementation of permessage-deflate, rather than a generic extension or list of extensions.

@agalakhov could you please check the comments that I left to this PR and check if I missed something? This is quite a huge PR and I would check it carefully before we can merge it. But I think there is some work that has to be done.

**Comment by SirCipher** [comment]
Hi @application-developer-DA, thank you for the review. Regarding the additional type parameter, the thought behind this was that it allowed for users to implement their own extensions and provide them. An alternative approach to this would be to have permessage-deflate configurable in the `WebSocketConfig` structure and then select the appropriate encoder when building the `WebSocket`. If this is something that you would prefer, then I'm happy to make the change, and then that would resolve a number of the comments that you have made. 

Regarding your error handling/into comments, the `?` operator doesn't work with the trait's associated type for the error - which is why I matched on the error and then did `Into::into`, though this could be tidied up with `map_err`.

File: src/protocol/mod.rs
Comment (SirCipher) [reply src/protocol/mod.rs:534]: The blocks that you've mentioned this on could be tidied up using `map_err`, but they don't work with `?` and the trait's associated type.

File: src/protocol/mod.rs
Comment (SirCipher) [reply src/protocol/mod.rs:439]: If the deflate extension is disabled, then it delegates the work to the uncompressed extension and does not mutate the frame in any way.

File: src/handshake/client.rs
Comment (SirCipher) [reply src/handshake/client.rs:71]: Upon the handshake finishing, the contents of the option are removed and the `WebSocket` instance is created. With some additional work, this could possibly be tidied up. Or, if the functionality to provide custom implementations of extensions is removed, then this will be removed.

File: src/extensions/uncompressed.rs
Comment (SirCipher) [reply src/extensions/uncompressed.rs:101]: It is rather tightly coupled with the caller in that regard. Perhaps the frame could be deconstructed into the required parts to remove the `unreachable` call.

File: src/extensions/deflate.rs
Comment (SirCipher) [reply src/extensions/deflate.rs:1]: It's an adaptation of WS-RS's deflate extension to use ZLib instead of libc, and to work with a growable fragment buffer.

**Comment by application-developer-DA** [comment]
@SirCipher , yes, I think having it inside a config makes sense. It seems like it would simplify many things. What do you think?

File: src/extensions/uncompressed.rs
Comment (application-developer-DA) [reply src/extensions/uncompressed.rs:101]: Yeah, that's how I imagined it.

File: src/extensions/deflate.rs
Comment (application-developer-DA) [reply src/extensions/deflate.rs:1]: Got it. I think I or @agalakhov will have to take a close look at it. At the first glance, it seems like it contains some copy-paste that we would like to avoid :)

**Comment by novacrazy** [comment]
What's the status of this PR?

**Comment by application-developer-DA** [comment]
@novacrazy the PR does not pass all of our CI/CD checks and has some conflicts, thus it can't be merged. I'm going to look at the content one more time to see if the rest is fine.

**Review by application-developer-DA — CHANGES_REQUESTED** [review]
Thanks for doing that great progress, but unfortunately there are several reasons why we can't merge it right now:

1. The PR does not pass our current CI/CD pipeline.
2. It looks like part of this PR should be done as a separate library, i.e. I think that the whole deflate/compression/uncompression logic here should not be part of `tungstenite-rs` as it could be generalized and does not have anything to do with WebSockets. This would make the implementation much more simple and concise. I'm afraid that if we merge it "as is" (even if (1) is fixed), we would introduce some technical debt to the library which we would like to avoid if possible, unless I'm missing on something.

**Comment by SirCipher** [comment]
Due to time constraints on my end and the additional changes required to make the requested changes, I won't be able to finish this PR in the near future.

**Pull request closed without merging**
